File: src/test/java/todayquest/user/controller/UserControllerUnitTest.java
package todayquest.user.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.mockito.MockedStatic;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.context.annotation.FilterType;
import org.springframework.core.io.ClassPathResource;
import org.springframework.core.io.ResourceLoader;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import todayquest.annotation.WithCustomMockUser;
import todayquest.common.MessageUtil;
import todayquest.config.SecurityConfig;
import todayquest.item.service.ItemLogService;
import todayquest.quest.service.QuestLogService;
import todayquest.user.dto.UserRequestDto;
import todayquest.user.service.UserService;

import java.util.Map;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.*;
import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@Slf4j
@DisplayName("유저 컨트롤러 유닛 테스트")
@WithCustomMockUser
@WebMvcTest(controllers = UserController.class,
        excludeFilters = @ComponentScan.Filter(
                type = FilterType.ASSIGNABLE_TYPE,
                classes = SecurityConfig.class
        )
)
public class UserControllerUnitTest {

    @Autowired
    MockMvc mvc;

    @MockBean
    ResourceLoader resourceLoader;

    @MockBean
    QuestLogService questLogService;

    @MockBean
    ItemLogService itemLogService;

    @MockBean
    UserService userService;

    static final String URI_PREFIX = "/user";

    @DisplayName("상태창 조회 테스트")
    @Test
    public void testMyPage() throws Exception {
        //given
        Long userId = 1L;
        String uri = "/status";
        Map<String, Long> questLog = Map.of(
                "COMPLETE", 10L,
                "FAIL",5L,
                "DISCARD", 3L);
        Map<String, Long> itemLog = Map.of(
                "EARN", 5L,
                "USE",2L,
                "ABANDON", 1L);

        when(resourceLoader.getResource(anyString())).thenReturn(new ClassPathResource("data/exp_table.json"));
        when(questLogService.getQuestLog(userId)).thenReturn(questLog);
        when(itemLogService.getItemLog(userId)).thenReturn(itemLog);


        //when
        //then
        mvc.perform(get(URI_PREFIX + uri))
                .andExpect(status().isOk())
                .andExpect(model().attribute("questLog", questLog))
                .andExpect(model().attribute("itemLog", itemLog))
                .andExpect(view().name("user/status"));

        verify(questLogService).getQuestLog(userId);
        verify(itemLogService).getItemLog(userId);
    }

    @DisplayName("닉네임 변경 API 테스트_성공")
    @Test
    public void testChangeNicknameApi() throws Exception {
        //given
        String uri = "";
        String nickname = "newNickname";
        UserRequestDto dto = new UserRequestDto();
        dto.setNickname(nickname);

        ObjectMapper om = new ObjectMapper();

        try (MockedStatic<MessageUtil> messageUtil = mockStatic(MessageUtil.class)) {
            when(userService.isDuplicateNickname(nickname)).thenReturn(false);
            when(MessageUtil.getMessage(anyString())).thenReturn("");

            //when
            mvc.perform(put(URI_PREFIX + uri)
                            .with(csrf())
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(om.writeValueAsString(dto)))
                    .andExpect(status().isOk())
                    .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                    .andExpect(content().string(MessageUtil.getMessage("nickname.changed")));
        }

        //then
        verify(userService).isDuplicateNickname(nickname);
        verify(userService).changeNickname(any(), eq(nickname));
    }

    @DisplayName("닉네임 변경 API 테스트_닉네임 중복")
    @Test
    public void testChangeNicknameApiDuplicate() throws Exception {
        //given
        String uri = "";
        String nickname = "newNickname";
        UserRequestDto dto = new UserRequestDto();
        dto.setNickname(nickname);

        ObjectMapper om = new ObjectMapper();

        try (MockedStatic<MessageUtil> messageUtil = mockStatic(MessageUtil.class)) {
            when(userService.isDuplicateNickname(nickname)).thenReturn(true);
            when(MessageUtil.getMessage(anyString())).thenReturn("");

            //when
            mvc.perform(put(URI_PREFIX + uri)
                            .with(csrf())
                            .contentType(MediaType.APPLICATION_JSON)
                            .content(om.writeValueAsString(dto)))
                    .andExpect(status().isConflict())
                    .andExpect(content().contentType(MediaType.APPLICATION_JSON))
                    .andExpect(content().string(MessageUtil.getMessage("nickname.changed")));
        }

        //then
        verify(userService).isDuplicateNickname(nickname);
        verifyNoMoreInteractions(userService);

    }



}

File: src/test/java/todayquest/user/controller/UserControllerUnitTest.kt
package todayquest.user.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule
import org.junit.jupiter.api.*
import org.junit.jupiter.api.extension.ExtensionContext
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.ArgumentsProvider
import org.junit.jupiter.params.provider.ArgumentsSource
import org.mockito.ArgumentMatchers
import org.mockito.MockedStatic
import org.mockito.Mockito
import org.mockito.kotlin.any
import org.mockito.kotlin.doThrow
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.boot.test.mock.mockito.MockBean
import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.FilterType
import org.springframework.dao.DataIntegrityViolationException
import org.springframework.http.MediaType
import org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.patch
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
import todayquest.annotation.WithCustomMockUser
import todayquest.common.MessageUtil
import todayquest.config.SecurityConfig
import todayquest.jwt.JwtAuthorizationFilter
import todayquest.user.dto.UserRequestDto
import todayquest.user.service.UserService
import java.util.stream.Stream

@DisplayName("유저 API 컨트롤러 유닛 테스트")
@WithCustomMockUser
@WebMvcTest(controllers = [UserApiController::class],
    excludeFilters = [
        ComponentScan.Filter(
            type = FilterType.ASSIGNABLE_TYPE,
            classes = [SecurityConfig::class, JwtAuthorizationFilter::class]
        )
    ]
)
class UserControllerUnitTest {

    companion object {
        const val URI_PREFIX = "/api/v1/users"
    }

    @Autowired
    lateinit var mvc: MockMvc

    @MockBean
    lateinit var userService: UserService

    private lateinit var messageUtil: MockedStatic<MessageUtil>

    val om: ObjectMapper = ObjectMapper().registerModule(JavaTimeModule())

    @BeforeEach
    fun init() {
        messageUtil = Mockito.mockStatic(MessageUtil::class.java)
        Mockito.`when`(MessageUtil.getMessage(ArgumentMatchers.anyString())).thenReturn("")
        Mockito.`when`(MessageUtil.getMessage(ArgumentMatchers.anyString(), ArgumentMatchers.any())).thenReturn("")
    }

    @AfterEach
    fun afterEach() {
        messageUtil.close()
    }

    class InValidUserSettingRequest: ArgumentsProvider {
        override fun provideArguments(context: ExtensionContext?): Stream<out Arguments> {
            val invalidPattern = UserRequestDto()
            invalidPattern.nickname = "ㄱㅂㄷㄴinvalidpattern"

            val invalidLength = UserRequestDto()
            invalidLength.nickname = "overnicknamelengthlimit123456789012345678901"

            val startWithWhiteSpace = UserRequestDto()
            startWithWhiteSpace.nickname = " startWhiteSpace"

            val endWithWhiteSpace = UserRequestDto()
            endWithWhiteSpace.nickname = "endWhiteSpace "

            val specialCharacter = UserRequestDto()
            specialCharacter.nickname = "contain_special_chracter"

            return Stream.of(
                Arguments.of(invalidPattern),
                Arguments.of(invalidLength),
                Arguments.of(startWithWhiteSpace),
                Arguments.of(endWithWhiteSpace),
                Arguments.of(specialCharacter),
            )
        }
    }

    @DisplayName("유저 설정 수정 시")
    @Nested
    inner class UserSettingsUpdateTest {

        @ArgumentsSource(InValidUserSettingRequest::class)
        @DisplayName("올바른 닉네임 패턴 요청이 아닌 경우 400이 반환된다")
        @ParameterizedTest(name = "{0} 값이 들어오면 400이 반환된다")
        fun `올바른 닉네임 패턴 요청이 아닌 경우 400이 반환된다`(dto: UserRequestDto) {
            //given

            //when
            val result = mvc.perform(
                patch(URI_PREFIX)
                    .content(om.writeValueAsBytes(dto))
                    .contentType(MediaType.APPLICATION_JSON_UTF8)
                    .with(csrf())
            )

            //then
            result
                .andExpect(status().isBadRequest)
        }

        @DisplayName("올바른 닉네임 패턴 요청일 경우 200이 반환된다")
        @Test
        fun `올바른 닉네임 패턴 요청일 경우 200이 반환된다`() {
            //given
            val dto = UserRequestDto()
            dto.nickname = "newNickname"

            //when
            val result = mvc.perform(
                patch(URI_PREFIX)
                    .content(om.writeValueAsBytes(dto))
                    .contentType(MediaType.APPLICATION_JSON_UTF8)
                    .with(csrf())
            )

            //then
            result
                .andExpect(status().isOk)
        }

        @DisplayName("닉네임이 null일 경우 200이 반환된다")
        @Test
        fun `닉네임이 null일 경우 200이 반환된다`() {
            //given
            val dto = UserRequestDto()

            //when
            val result = mvc.perform(
                patch(URI_PREFIX)
                    .content(om.writeValueAsBytes(dto))
                    .contentType(MediaType.APPLICATION_JSON_UTF8)
                    .with(csrf())
            )

            //then
            result
                .andExpect(status().isOk)
        }

        @DisplayName("제약조건 위반 예외 발생 시 409를 반환한다")
        @Test
        fun `제약조건 위반 예외 발생 시 409를 반환한다`() {
            //given
            val dto = UserRequestDto()
            doThrow(DataIntegrityViolationException::class).`when`(userService).changeUserSettings(any(), any())

            //when
            val result = mvc.perform(
                patch(URI_PREFIX)
                    .content(om.writeValueAsBytes(dto))
                    .contentType(MediaType.APPLICATION_JSON_UTF8)
                    .with(csrf())
            )

            //then
            result
                .andExpect(status().isConflict)
        }
    }


}